src/native/marker-index.cc

```diff
@@ -0,0 +1,420 @@
+        end_node->left_marker_ids.erase(id);
+      }
+      end_nodes_by_id[id] = start_node;
+    }
+    DeleteNode(end_node);
+  } else if (end_node->IsMarkerEndpoint()) {
+    end_node->priority = GenerateRandomNumber();
+    BubbleNodeDown(end_node);
+  } else {
+    DeleteNode(end_node);
+  }
+
+  if (start_node->IsMarkerEndpoint()) {
+    start_node->priority = GenerateRandomNumber();
+    BubbleNodeDown(start_node);
+  } else {
+    DeleteNode(start_node);
+  }
+
+  return invalidated;
+}
+
+Point MarkerIndex::GetStart(MarkerId id) const {
+  auto result = start_nodes_by_id.find(id);
+  if (result == start_nodes_by_id.end())
+    return Point();
+  else
+    return GetNodeOffset(result->second);
+}
+
+Point MarkerIndex::GetEnd(MarkerId id) const {
+  auto result = end_nodes_by_id.find(id);
+  if (result == end_nodes_by_id.end())
+    return Point();
+  else
+    return GetNodeOffset(result->second);
+}
+
+unordered_set<MarkerId> MarkerIndex::FindIntersecting(Point start, Point end) {
+  unordered_set<MarkerId> result;
+  iterator.FindIntersecting(start, end, &result);
+  return result;
+}
+
+unordered_set<MarkerId> MarkerIndex::FindContaining(Point start, Point end) {
+  unordered_set<MarkerId> containing_start;
+  iterator.FindIntersecting(start, start, &containing_start);
+  if (end == start) {
+    return containing_start;
+  } else {
+    unordered_set<MarkerId> containing_end;
+    iterator.FindIntersecting(end, end, &containing_end);
+    unordered_set<MarkerId> containing_start_and_end;
+    for (const MarkerId &id : containing_start) {
+      if (containing_end.count(id) > 0) {
+        containing_start_and_end.insert(id);
+      }
+    }
+    return containing_start_and_end;
+  }
+}
+
+unordered_set<MarkerId> MarkerIndex::FindContainedIn(Point start, Point end) {
+  unordered_set<MarkerId> result;
+  iterator.FindContainedIn(start, end, &result);
+  return result;
+}
+
+unordered_set<MarkerId> MarkerIndex::FindStartingIn(Point start, Point end) {
+  unordered_set<MarkerId> result;
+  iterator.FindStartingIn(start, end, &result);
+  return result;
+}
+
+unordered_set<MarkerId> MarkerIndex::FindEndingIn(Point start, Point end) {
+  unordered_set<MarkerId> result;
+  iterator.FindEndingIn(start, end, &result);
+  return result;
+}
+
+unordered_map<MarkerId, Range> MarkerIndex::Dump() {
+  return iterator.Dump();
+}
+
+Point MarkerIndex::GetNodeOffset(const Node *node) const {
+  Point offset = node->left_extent;
+  while (node->parent) {
+    if (node->parent->right == node) {
+      offset = node->parent->left_extent.Traverse(offset);
+    }
+
+    node = node->parent;
+  }
+  return offset;
+}
+
+void MarkerIndex::DeleteNode(Node *node) {
+  node->priority = INT_MAX;
+
+  BubbleNodeDown(node);
+
+  if (node->parent) {
+    if (node->parent->left == node) {
+      node->parent->left = nullptr;
+    } else {
+      node->parent->right = nullptr;
+    }
+  } else {
+    root = nullptr;
+  }
+
+  delete node;
```

Comment:
Yeah, really good call about not using smart pointers here. This reads perfectly.